### PR TITLE
NR-250703: cluster-autoscaler improvements for GCP

### PR DESCRIPTION
**Pull request by sachin-shankar**
Add support for label(similar to tags in Azure) based auto-discovery of GCP Managed Instance Groups. The discovery includes figuring out the min and max sizes for the mig pool via labels as well.

Add unit-tests for all the new code added.

Add GCP auto-discovery documentation.

#### Which issue(s) this PR fixes:

Fixes #

#### Does this PR introduce a user-facing change?

```release-note

```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

```docs

```

### CODE REVIEW

File: cluster-autoscaler/Makefile [hunk 1]
```diff
@@ -1,13 +1,13 @@
...
 
-TAG?=dev
+TAG?=v1.25.3_NR
 FLAGS=
 LDFLAGS?=-s
 ENVVAR=CGO_ENABLED=0
 GOOS?=linux
 GOARCH?=$(shell go env GOARCH)
-REGISTRY?=staging-k8s.gcr.io
+REGISTRY?=cf-registry.nr-ops.net/container-fabric
```
Comment (Bryce-Soghigian) [inline cluster-autoscaler/Makefile:10]: ?!!

File: cluster-autoscaler/cloudprovider/gce/gce_manager.go [hunk 2]
```diff
@@ -614,14 +659,30 @@ func (m *gceManagerImpl) GetMigTemplateNode(mig Mig) (*apiv1.Node, error) {
...
-	for i, spec := range o.NodeGroupAutoDiscoverySpecs {
-		cfgs[i], err = parseMIGAutoDiscoverySpec(spec)
+	var cfgs []migAutoDiscoveryConfig
+	var labelEntryFound bool
+	for _, spec := range o.NodeGroupAutoDiscoverySpecs {
+		c, err := parseMIGAutoDiscoverySpec(spec)
 		if err != nil {
 			return nil, err
+		} else if labelEntryFound {
+			if c.Labels != nil {
```
Comment (x13n) [inline cluster-autoscaler/cloudprovider/gce/gce_manager.go:673]: Why is this an error?

File: cluster-autoscaler/cloudprovider/gce/gce_manager.go [hunk 3]
```diff
@@ -614,14 +659,30 @@ func (m *gceManagerImpl) GetMigTemplateNode(mig Mig) (*apiv1.Node, error) {
...
+// returns an error if more than 1 label-based entry is specified
 func parseMIGAutoDiscoverySpecs(o cloudprovider.NodeGroupDiscoveryOptions) ([]migAutoDiscoveryConfig, error) {
-	cfgs := make([]migAutoDiscoveryConfig, len(o.NodeGroupAutoDiscoverySpecs))
-	var err error
-	for i, spec := range o.NodeGroupAutoDiscoverySpecs {
-		cfgs[i], err = parseMIGAutoDiscoverySpec(spec)
+	var cfgs []migAutoDiscoveryConfig
+	var labelEntryFound bool
+	for _, spec := range o.NodeGroupAutoDiscoverySpecs {
+		c, err := parseMIGAutoDiscoverySpec(spec)
```
Comment (x13n) [inline cluster-autoscaler/cloudprovider/gce/gce_manager.go:669]: nit, optional: I think you could simplify the error handling here a bit by first determining the discoverer per spec and then comparing discoverer counts.

File: cluster-autoscaler/cloudprovider/gce/gce_manager.go [hunk 4]
```diff
@@ -634,21 +695,68 @@ type migAutoDiscoveryConfig struct {
...
+
+	if len(labels) == 0 {
+		return cfg, fmt.Errorf("no labels specified. use key1=value1,key2=value2")
 	}
+	cfg.Labels = labels
+	cfg.Re = regexp.MustCompile(".*") // match all
 
-	for _, arg := range strings.Split(tokens[1], ",") {
+	// set default min, max size. will be overridden by instance template labels if present
+	if scaleToZeroSupported {
```
Comment (x13n) [inline cluster-autoscaler/cloudprovider/gce/gce_manager.go:743]: nit: This is a const that has been set to true for past 7 years, I highly doubt we'll stop supporting scale to 0 at this point.

File: cluster-autoscaler/cloudprovider/gce/gce_manager_test.go [hunk 5]
```diff
@@ -1249,6 +1251,167 @@ func TestFetchAutoMigsRegional(t *testing.T) {
 	mock.AssertExpectationsForObjects(t, server)
 }
 
+func TestFetchAutoMigsWithMatchingLabelsInInstanceTemplates(t *testing.T) {
```
Comment (x13n) [inline cluster-autoscaler/cloudprovider/gce/gce_manager_test.go:1254]: Can you make these tests table-based?

File: cluster-autoscaler/cloudprovider/gce/gce_manager_test.go [hunk 6]
```diff
@@ -1470,6 +1657,151 @@ func TestParseMIGAutoDiscoverySpecs(t *testing.T) {
 	}
 }
 
+func TestParseMIGAutoDiscoverySpec(t *testing.T) {
```
Comment (x13n) [inline cluster-autoscaler/cloudprovider/gce/gce_manager_test.go:1660]: Why is this (and the following 2 tests) not simply a part of `TestParseMIGAutoDiscoverySpecs`?

**Review by x13n — CHANGES_REQUESTED** [review]

**Comment by x13n** [comment]
This doesn't seem to be worked on, so closing for now. @sachin-shankar Please reopen when you have the time to follow up. 

/close

**Pull request closed without merging**